File: auth-center/auth-center-domain/src/test/java/com/jmsoftware/maf/authcenter/security/service/impl/JwtServiceImplTest.java
package com.jmsoftware.maf.authcenter.security.service.impl;

import com.google.common.collect.Lists;
import com.jmsoftware.maf.common.exception.SecurityException;
import com.jmsoftware.maf.springcloudstarter.property.JwtConfigurationProperties;
import io.jsonwebtoken.Claims;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ValueOperations;

import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

/**
 * <h1>JwtServiceImplTest</h1>
 * Description: JwtServiceImplTest, change description here.
 * <p>
 * <h2>Mockito JUnit 5 Extension</h2>
 * <p>There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.</p>
 * <p><strong>Pros:</strong></p>
 * <ul>
 * <li>No need to call <code>MockitoAnnotations.openMocks()</code></li>
 * <li>Validates framework usage and detects incorrect stubbing</li>
 * <li>Easy to create mocks</li>
 * <li>Very readable</li>
 *
 * </ul>
 * <p><strong>Cons:</strong></p>
 * <ul>
 * <li>Need an extra dependency on <code>org.mockito:mockito-junit-jupiter</code>, which has been included by Spring.
 * So we don&#39;t have to worry about this.</li>
 * </ul>
 *
 * @author Johnny Miller (鍾俊), e-mail: johnnysviva@outlook.com, date: 4/4/2022 8:40 AM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito With JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito With JUnit 5</a>
 **/
@Slf4j
@ExtendWith(MockitoExtension.class)
@Execution(ExecutionMode.CONCURRENT)
class JwtServiceImplTest {
    private final static long USER_ID = 1L;
    private final static String USERNAME = "ijohnnymiller";
    private final static String JWT = "eyJhbGciOiJIUzM4NCJ9" +
            ".eyJqdGkiOiIxIiwic3ViIjoiaWpvaG5ueW1pbGxlciIsImlhdCI6MTY0OTA0NTgzMiwicm9sZXMiOltdfQ" +
            ".lMQDOTnNsZTaopDBf7O_4zlhPJTZmGB3v6MpjmWnPuMEb2Aj-bXciDh4m2pAYtEp";
    @Spy
    @InjectMocks
    private JwtServiceImpl jwtService;
    @Mock
    private JwtConfigurationProperties jwtConfigurationProperties;
    @Mock
    private RedisTemplate<Object, Object> redisTemplate;
    @Mock
    private ValueOperations<Object, Object> valueOperations;

    @BeforeEach
    void setUp() {
        log.info("{} setUp", this.getClass().getSimpleName());
        when(this.jwtConfigurationProperties.getSigningKey())
                .thenReturn("signing-key-for-unit-test-only-do-not-use-in-production");
        this.jwtService.init();
    }

    @AfterEach
    void tearDown() {
        log.info("{} tearDown", this.getClass().getSimpleName());
    }

    @Test
    void createJwt() {
        when(this.redisTemplate.opsForValue()).thenReturn(this.valueOperations);
        doNothing().when(this.valueOperations).set(anyString(), anyString(), anyLong(), any());
        val jwt = this.jwtService.createJwt(true, USER_ID, USERNAME, Lists.newArrayList(), Lists.newArrayList());
        assertNotEquals(0, jwt.length());
    }

    @Test
    void parseJwt() {
        when(this.redisTemplate.opsForValue()).thenReturn(this.valueOperations);
        when(this.valueOperations.getOperations()).thenReturn(this.redisTemplate);
        val claimsAtomicReference = new AtomicReference<Claims>();
        assertThrows(SecurityException.class, () -> claimsAtomicReference.set(this.jwtService.parseJwt(JWT)));
        log.info("Parse JWT. claimsAtomicReference: {}", claimsAtomicReference.hashCode());
        assertNotNull(claimsAtomicReference);
        assertNull(claimsAtomicReference.get());
    }

    @Test
    void invalidateJwt() {
        when(this.redisTemplate.opsForValue()).thenReturn(this.valueOperations);
        when(this.valueOperations.getOperations()).thenReturn(this.redisTemplate);
        doReturn(JWT).when(this.jwtService).getJwtFromRequest(any());
        assertThrows(SecurityException.class, () -> this.jwtService.invalidateJwt(any()));
    }

    @Test
    void getUsernameFromJwt() {
        when(this.redisTemplate.opsForValue()).thenReturn(this.valueOperations);
        when(this.valueOperations.getOperations()).thenReturn(this.redisTemplate);
        assertThrows(SecurityException.class, () -> this.jwtService.getUsernameFromJwt(JWT));
    }

    @Test
    void getUsernameFromRequest() {
        when(this.redisTemplate.opsForValue()).thenReturn(this.valueOperations);
        when(this.valueOperations.getOperations()).thenReturn(this.redisTemplate);
        doReturn(JWT).when(this.jwtService).getJwtFromRequest(any());
        assertThrows(SecurityException.class, () -> this.jwtService.getUsernameFromRequest(any()));
    }

    @Test
    void getJwtFromRequest() {
        assertThrows(NullPointerException.class, () -> this.jwtService.getJwtFromRequest(any()));
    }

    @Test
    void parse() {
        assertThrows(NullPointerException.class, () -> this.jwtService.parse(any()));
    }
}

File: auth-center/auth-center-domain/src/test/kotlin/com/jmsoftware/maf/authcenter/security/service/impl/JwtServiceImplTest.kt
package com.jmsoftware.maf.authcenter.security.service.impl

import cn.hutool.core.util.StrUtil
import com.jmsoftware.maf.common.exception.SecurityException
import com.jmsoftware.maf.common.util.logger
import com.jmsoftware.maf.springcloudstarter.property.JwtConfigurationProperties
import io.jsonwebtoken.Claims
import jakarta.servlet.http.HttpServletRequest
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.Spy
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.ValueOperations
import java.util.concurrent.atomic.AtomicReference


/**
 * # JwtServiceImplTest
 *
 * Description: JwtServiceImplTest, change description here.
 *
 * ## Mockito JUnit 5 Extension
 *
 * There is also a Mockito extension for JUnit 5 that will make the initialization even simpler.
 *
 * **Pros:**
 *
 *  * No need to call `MockitoAnnotations.openMocks()`
 *  * Validates framework usage and detects incorrect stubbing
 *  * Easy to create mocks
 *  * Very readable
 *
 * **Cons:**
 *
 *  * Need an extra dependency on `org.mockito:mockito-junit-jupiter`, which has been included by Spring.
 * So we don&#39;t have to worry about this.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/11/22 9:29 PM
 * @see <a href='https://www.arhohuttunen.com/junit-5-mockito/'>Using Mockito with JUnit 5</a>
 * @see <a href='https://www.youtube.com/watch?v=p7_cTAF39A8/'>YouTube - Using Mockito with JUnit 5</a>
 */
@ExtendWith(MockitoExtension::class)
class JwtServiceImplTest {
    companion object {
        private val log = logger()
        private const val USER_ID = 1L
        private const val USERNAME = "ijohnnymiller"
        private const val JWT = "eyJhbGciOiJIUzM4NCJ9" +
                ".eyJqdGkiOiIxIiwic3ViIjoiaWpvaG5ueW1pbGxlciIsImlhdCI6MTY0OTA0NTgzMiwicm9sZXMiOltdfQ" +
                ".lMQDOTnNsZTaopDBf7O_4zlhPJTZmGB3v6MpjmWnPuMEb2Aj-bXciDh4m2pAYtEp"
    }

    @Spy
    @InjectMocks
    private lateinit var jwtService: JwtServiceImpl

    @Mock
    private lateinit var jwtConfigurationProperties: JwtConfigurationProperties

    @Mock
    private lateinit var redisTemplate: RedisTemplate<Any, Any>

    @Mock
    private lateinit var valueOperations: ValueOperations<Any, Any>

    @Mock
    private lateinit var httpServletRequest: HttpServletRequest

    @BeforeEach
    fun setUp() {
        log.info("${this.javaClass.simpleName} setUp")
        `when`(jwtConfigurationProperties.signingKey)
            .thenReturn("signing-key-for-unit-test-only-do-not-use-in-production")
        jwtService.init()
    }

    @AfterEach
    fun tearDown() {
        log.info("${this.javaClass.simpleName} tearDown")
    }

    @Test
    fun createJwt() {
        `when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        doNothing().`when`(valueOperations)[anyString(), anyString(), anyLong()] = any()
        val jwt = jwtService.createJwt(true, USER_ID, USERNAME, listOf(), listOf())
        assertNotEquals(0, jwt.length)
    }

    @Test
    fun parseJwt() {
        `when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        `when`(valueOperations.operations).thenReturn(redisTemplate)
        val claimsAtomicReference = AtomicReference<Claims>()
        assertThrows(SecurityException::class.java) { claimsAtomicReference.set(jwtService.parseJwt(JWT)) }
        log.info("Parse JWT. claimsAtomicReference: ${claimsAtomicReference.hashCode()}")
        assertNotNull(claimsAtomicReference)
        assertNull(claimsAtomicReference.get())
    }

    @Test
    fun invalidateJwt() {
        `when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        `when`(valueOperations.operations).thenReturn(redisTemplate)
        doReturn(JWT).`when`(jwtService).getJwtFromRequest(httpServletRequest)
        assertThrows(SecurityException::class.java) { jwtService.invalidateJwt(httpServletRequest) }
    }

    @Test
    fun usernameFromJwt() {
        `when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        `when`(valueOperations.operations).thenReturn(redisTemplate)
        assertThrows(SecurityException::class.java) { jwtService.getUsernameFromJwt(JWT) }
    }

    @Test
    fun getUsernameFromRequest() {
        `when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        `when`(valueOperations.operations).thenReturn(redisTemplate)
        doReturn(JWT).`when`(jwtService).getJwtFromRequest(httpServletRequest)
        val thrownException =
            assertThrows(SecurityException::class.java) { jwtService.getUsernameFromRequest(httpServletRequest) }
        log.warn("Thrown exception: ${thrownException.message}")
    }

    @Test
    fun jwtFromRequest() {
        val jwt = jwtService.getJwtFromRequest(httpServletRequest)
        log.info("JWT: $jwt")
        assertTrue(StrUtil.isNotBlank(jwt))
    }

    @Test
    fun parse() {
        val thrownException = assertThrows(SecurityException::class.java) { jwtService.parse(httpServletRequest) }
        log.warn("Thrown exception: ${thrownException.message}")
    }
}
